dynamax/utils.py

```diff
@@ -87,4 +84,34 @@ def random_rotation(seed, n, theta=None):
     out = jnp.eye(n)
     out = out.at[:2, :2].set(rot)
     q = jnp.linalg.qr(jr.uniform(key2, shape=(n, n)))[0]
-    return q.dot(out).dot(q.T)
\ No newline at end of file
+    return q.dot(out).dot(q.T)
+
+
+def add_batch_dim(tree, instance_shapes):
+    """Add a batch dimension to an array, if necessary.
+
+    Args:
+        tree (_type_): PyTree whose leaves' shapes are either
+            (batch, length) + instance_shape or (length,) + instance_shape.
+            If the latter, this function adds a batch dimension of 1 to
+            each leaf node.
+
+        instance_shape (_type_): matching PyTree with tuple of integers specifying the shape
+            of one "instance" or entry in the array. where (-1)
```

Comment:
unfinished comment about -1

Reply:
It may help to list a couple of examples of this function in action (input / output shapes).

Reply:
Good call. I simplified this function since I wasn't using the `(-1)` functionality anyway. I also added some examples.